File: RPGFgin/src/core/mesh.ts
import { v4 as uuidv4 } from "uuid";

import { gl } from "../main";
import { StaticVerticesBuffer } from "./staticVerticesBuffer";
import { StaticIndexBuffer } from "./staticIndexBuffer";
import { mat4 } from "glm-js";
import { Shader } from "./shader";
import { Renderer } from "./renderer";
import { ITexture } from "../types/ITexture";

export enum TextureTypes {
  DIFFUSE = "texture_diffuse",
  SPECULAR = "texture_specular",
  NORMAL = "texture_normal",
  HEIGHT = "texture_height",
  CUBE = "cube",
}

export const TextureSlotsMap = {
  [TextureTypes.DIFFUSE]: 0,
  [TextureTypes.SPECULAR]: 1,
  [TextureTypes.NORMAL]: 2,
  [TextureTypes.HEIGHT]: 3,
  [TextureTypes.CUBE]: 4,
};
export class Mesh {
  numIndices: number;
  VAO: WebGLVertexArrayObject;
  VBO: StaticVerticesBuffer;
  TBO: StaticVerticesBuffer;
  NBO: StaticVerticesBuffer;
  EBO: StaticIndexBuffer;
  allowIntersections: boolean = true;
  uuid: string;
  modelMatrix = new mat4();

  constructor(
    public vertices: Float32Array,
    public indices: Uint32Array,
    public texCoords: Float32Array = new Float32Array([]),
    public normals: Float32Array = new Float32Array([]),
    public textures: ITexture[] = []
  ) {
    this.uuid = uuidv4();
    this.numIndices = indices.length;

    this.VAO = gl.createVertexArray();
    gl.bindVertexArray(this.VAO);
    this.VBO = new StaticVerticesBuffer(vertices);

    gl.enableVertexAttribArray(0);
    gl.vertexAttribPointer(0, 3, gl.FLOAT, false, 0, null);

    if (normals.length) {
      this.NBO = new StaticVerticesBuffer(normals);
      gl.enableVertexAttribArray(1);
      gl.vertexAttribPointer(1, 3, gl.FLOAT, false, 0, null);
    }

    this.TBO = new StaticVerticesBuffer(texCoords);
    gl.enableVertexAttribArray(2);
    gl.vertexAttribPointer(2, 2, gl.FLOAT, false, 0, null);

    this.EBO = new StaticIndexBuffer(indices);

    gl.bindVertexArray(null);
  }

  getTextures(): ITexture[] {
    return this.textures;
  }

  getModelMatrix(): mat4 {
    return this.modelMatrix;
  }

  render(shader: Shader, renderMode: number): void {
    if (!Renderer.activeMeshesMap[this.uuid] && this.allowIntersections) {
      return;
    }

    this.textures.forEach((texture) => {
      shader.setUniform1i("mainSampler", TextureSlotsMap[texture.type]);
    });

    gl.bindVertexArray(this.VAO);
    gl.drawElements(renderMode, this.numIndices, gl.UNSIGNED_INT, 0);
    gl.activeTexture(gl.TEXTURE0);
  }
}

File: src/Mesh.rs
// use crate::GL;
// use web_sys::{WebGl2RenderingContext, WebGlVertexArrayObject};

// use crate::core::{IndexBuffer::IndexBuffer, Shader::Shader, StaticBuffer::StaticBuffer};

// pub struct Mesh {
//     VBO: StaticBuffer,
//     IBO: IndexBuffer,
//     VAO: WebGlVertexArrayObject,
//     numberOfIndices: usize,
// }

// impl Mesh {
//     pub fn new(
//         vertices: Vec<f32>,
//         indices: Vec<u32>,
//         texCoords: Vec<f32>,
//         normals: Vec<f32>,
//     ) -> Mesh {
//         let gl = GL.context.borrow();
//         let VAO = gl.create_vertex_array().unwrap();
//         gl.bind_vertex_array(Some(&VAO));

//         let VBO = StaticBuffer::new(vertices, WebGl2RenderingContext::STATIC_DRAW);

//         gl.enable_vertex_attrib_array(0);
//         gl.vertex_attrib_pointer_with_i32(0, 3, WebGl2RenderingContext::FLOAT, false, 0, 0);

//         // gl.enable_vertex_attrib_array(1);
//         // gl.vertex_attrib_pointer_with_i32(1, 3, WebGl2RenderingContext::FLOAT, false, 0, 0);

//         // gl.enable_vertex_attrib_array(2);
//         // gl.vertex_attrib_pointer_with_i32(2, 2, WebGl2RenderingContext::FLOAT, false, 0, 0);

//         let numberOfIndices = indices.len();
//         let IBO = IndexBuffer::new(indices, WebGl2RenderingContext::STATIC_DRAW);

//         Mesh {
//             numberOfIndices,
//             VBO,
//             IBO,
//             VAO,
//         }
//     }
//     pub fn render(self, shader: &Shader) {
//         let gl = GL.context.borrow();
//         gl.bind_vertex_array(Some(&self.VAO));

//         gl.draw_elements_with_i32(
//             WebGl2RenderingContext::TRIANGLES,
//             self.numberOfIndices as i32,
//             WebGl2RenderingContext::UNSIGNED_INT,
//             0,
//         )
//     }
// }
